beacon-chain/blockchain/van_minimal_consensus_info.go

```diff
@@ -0,0 +1,197 @@
+)
+
+// MinimalConsensusInfoFetcher retrieves the minimal consensus info for provided epoch from blockchain service
+type MinimalConsensusInfoFetcher interface {
+	MinimalConsensusInfo(types.Epoch) (*ethpb.MinimalConsensusInfo, error)
+	MinimalConsensusInfoRange(types.Epoch) ([]*ethpb.MinimalConsensusInfo, error)
+}
+
+//MinimalConsensusInfo provides minimalConsensusInfo for one epoch provided as an argument
+func (s *Service) MinimalConsensusInfo(epoch types.Epoch) (minConsensusInfo *ethpb.MinimalConsensusInfo, err error) {
+	log.WithField("prefix", "GetPastMinimalConsensusInfo").WithField("epoch", uint64(epoch))
+
+	assignments, err := s.getPastProposerListForEpoch(epoch)
+	if nil != err {
+		log.Errorf("[VAN_SUB] getProposerListForEpoch err = %s", err.Error())
+		return nil, err
+	}
+
+	assignmentsSlice := make([]string, 0)
+
+	// Slot 0 was never signed by anybody
+	if 0 == epoch {
+		publicKeyBytes := make([]byte, params.BeaconConfig().BLSPubkeyLength)
+		currentString := fmt.Sprintf("0x%s", hex.EncodeToString(publicKeyBytes))
+		assignmentsSlice = append(assignmentsSlice, currentString)
+	}
+
+	for _, assigment := range assignments.Assignments {
+		currentString := fmt.Sprintf("0x%s", hex.EncodeToString(assigment.PublicKey))
+		assignmentsSlice = append(assignmentsSlice, currentString)
+	}
+
+	expectedValidators := int(params.BeaconConfig().SlotsPerEpoch)
+
+	if len(assignmentsSlice) != expectedValidators {
+		err := fmt.Errorf(
+			"not enough assignments, expected: %d, got: %d",
+			expectedValidators,
+			len(assignmentsSlice),
+		)
+		log.Errorf("[VAN_SUB] Assignments err = %s", err.Error())
+
+		return nil, err
+	}
+
+	genesisTime := s.genesisTime
+	startSlot, err := helpers.StartSlot(epoch)
+	if nil != err {
+		log.Errorf("[VAN_SUB] StartSlot err = %s", err.Error())
+		return nil, err
+	}
+	epochStartTime, err := helpers.SlotToTime(uint64(genesisTime.Unix()), startSlot)
+	if nil != err {
+		log.Errorf("[VAN_SUB] SlotToTime err = %s", err.Error())
+		return nil, err
+	}
+
+	minConsensusInfo = &ethpb.MinimalConsensusInfo{
+		Epoch:            epoch,
+		ValidatorList:    assignmentsSlice,
+		EpochTimeStart:   uint64(epochStartTime.Unix()),
+		SlotTimeDuration: &types2.Duration{Seconds: int64(params.BeaconConfig().SecondsPerSlot)},
+	}
+
+	log.Infof("[VAN_SUB] currEpoch = %#v", uint64(epoch))
+
+	return minConsensusInfo, nil
+}
+
+//MinimalConsensusInfoRange provides minimalConsensusInfo for all epochs from epoch provided as an argument
+func (s *Service) MinimalConsensusInfoRange(
+	fromEpoch types.Epoch,
+) (consensusInfos []*ethpb.MinimalConsensusInfo, err error) {
+	consensusInfo, err := s.MinimalConsensusInfo(fromEpoch)
+
+	if nil != err {
+		log.WithField("currentEpoch", "unknown").
+			WithField("requestedEpoch", fromEpoch).Error(err.Error())
+
+		return nil, err
+	}
+
+	consensusInfos = make([]*ethpb.MinimalConsensusInfo, 0)
+	consensusInfos = append(consensusInfos, consensusInfo)
+	tempEpochIndex := consensusInfo.Epoch
+
+	for {
```

Comment:
you can retrieve `currentEpoch := helpers.SlotToEpoch(bs.GenesisTimeFetcher.CurrentSlot())` and then iterate from `fromEpoch` to `currentEpoch` and prepare `consensusInfo` for each of the slot. I would not prefer using infinite loop in here.

Reply:
As we agreed on, we're leaving infinity loop because of gRPC streaming - need to have this loop to handle stream/connection.